lms/security.py

```diff
@@ -31,31 +31,35 @@ class Permissions(Enum):
 class SecurityPolicy:
     """Top-level authentication policy that delegates to sub-policies."""
 
-    def __init__(self):
-        self._subpolicies = [LTISecurityPolicy(), LMSGoogleSecurityPolicy()]
+    def __init__(self, path_policy_mapping, fallback_policy):
+        # Map between route paths and policies
+        self.path_policy_mapping = path_policy_mapping
```

Comment:
Injecting the configuration instead of hard-coding it. I reckon this makes testing easiser.

Reply:
I think you probably want to hard-code the mapping here because the mapping itself is a likely source of bugs and it isn't covered by the tests currently (I don't think the tests for `includeme()` are likely to catch bugs).

For example if I created a mapping like this:

```python
path_policy_mapping={
    "/admin": LMSGoogleSecurityPolicy,
    "/admin/foo": FooPolicy,
    "/googleauth": LMSGoogleSecurityPolicy,
    "/": LTISecurityPolicy,
    "/bar": BarPolicy,
}
```

`FooPolicy` and `BarPolicy` will never be called because of the ordering, but the unit tests won't catch this since they don't test with the actual mapping.

If the mapping were hard-coded then I'd find the mistake when I tried to add a `PathPolicyGetter` unit test like this:

```python
class TestPathPolicyGetter:
    ...

    # This test would fail.
    def test_it_returns_the_foo_policy_for_foo_pages(self, request):
        request.path = "/admin/foo/baz"

        assert isinstance(PathPolicyGetter()(request), FooPolicy)
```

`TestPathPolicyGetter` could mock the sub-policies with `patch` fixtures. That would probably be ideal as it means the real policies aren't used at all in `PathPolicyGetter`'s unit tests so bugs in the policies (for example in their `__init__()` methods) can't cause `PathPolicyGetter`'s tests to fail. With patching the assertions would look something like:

```python
assert PathPolicyGetter()(request) == FooPolicy.return_value
```

Normally I would say that unit tests should always mock out any other units _from our own code_ that the unit-under-test uses. But in this case `PathPolicyGetter` and the security policies are close-collaborators (they're in the same file) so I don't think the tests _have to_ mock. But I think the tests might actually be better if they _do_ mock.